src/build.cc

```diff
@@ -649,11 +670,17 @@ bool Builder::Build(string* err) {
           }
         } else {
           ++pending_commands;
-        }
 
-        // We made some progress; go back to the main loop.
-        continue;
+          --capacity;
+
+          // Re-evaluate capacity.
+          size_t current_capacity = command_runner_->CanRunMore();
+          if (current_capacity < capacity)
+            capacity = current_capacity;
+        }
       }
+
+      if (!plan_.more_to_do()) break;
```

Comment:
isn't this redundant?

Reply:
No, otherwise you may end up in the `ninja: build stopped: stuck [this is a bug].` case at the end of a build. However, I wonder if this should be

`if (pending_commands == 0 && !plan_more_to_do()) break;`

Reply:
Yep, otherwise we may no reap the last few finished commands, fixed that.